Manual/RecursiveDefs/WF/PreprocessExample.lean

```diff
@@ -0,0 +1,185 @@
+    sizeOf x < sizeOf p := by
+  cases h <;> cases p <;> (simp; omega)
+```
+
+The next step is to define {name Pair.attach}`attach` and {name Pair.unattach}`unattach` functions that enrich the elements of the pair with a proof that they are elements of the pair, or remove said proof.
+Here, the type of {name}`Pair.unattach` is more general and can be used with any {ref "Subtype"}[subtype]; this is a typical pattern.
+
+```lean
+def Pair.attach (p : Pair α) : Pair {x : α // x ∈ p} where
+  fst := ⟨p.fst, .fst⟩
+  snd := ⟨p.snd, .snd⟩
+
+def Pair.unattach {P : α → Prop} :
+    Pair {x : α // P x} → Pair α :=
+  Pair.map Subtype.val
+```
+
+{name Tree.map}`Tree.map` can now be defined by using {name}`Pair.attach` and {name}`Pair.sizeOf_lt_of_mem` explicitly:
+
+```lean (keep := false)
+def Tree.map (f : α → β) : Tree α → Tree β
+  | leaf x => leaf (f x)
+  | node p => node (p.attach.map (fun ⟨t', _⟩ => t'.map f))
+termination_by t => t
+decreasing_by
+  have := Pair.sizeOf_lt_of_mem ‹_›
+  simp_all +arith
+  omega
+```
+
+This transformation can be made fully automatic.
+The preprocessing feature of well-founded recursion can be used to automate the introduction of the {lean}`Pair.attach` function.
+This is done in two stages.
+First, when {name}`Pair.map` is applied to one of the function's parameters, it is rewritten to an {name Pair.attach}`attach`/{name Pair.unattach}`unattach` composition.
+Then, when a function is mapped over the result of {name}`Pair.unattach`, the function is rewritten to accept the proof of membership and bring it into scope.
+```lean
+@[wf_preprocess]
+theorem Pair.map_wfParam (f : α → β) (p : Pair α) :
+    (wfParam p).map f = p.attach.unattach.map f := by
+  cases p
+  simp [wfParam, Pair.attach, Pair.unattach, Pair.map]
+
+@[wf_preprocess]
+theorem Pair.map_unattach {P : α → Prop}
+    (p : Pair (Subtype P)) (f : α → β) :
+    p.unattach.map f =
+    p.map fun ⟨x, h⟩ =>
+      binderNameHint x f <|
+      f (wfParam x) := by
+  cases p; simp [wfParam, Pair.unattach, Pair.map]
+```
+
+Now the function body can be written without extra considerations, and the membership assumption is still available to the termination proof.
+
+```lean (keep := false)
+def Tree.map (f : α → β) : Tree α → Tree β
+  | leaf x => leaf (f x)
+  | node p => node (p.map (fun t' => t'.map f))
+termination_by t => t
+decreasing_by
+  have := Pair.sizeOf_lt_of_mem ‹_›
+  simp_all
+  omega
+```
+
+The proof can be made fully automatic by adding {name Pair.sizeOf_lt_of_mem}`sizeOf_lt_of_mem` to the {tactic}`decreasing_trivial` tactic, as is done for similar built-in theorems.
+
+```lean
+macro "sizeOf_pair_dec" : tactic =>
+  `(tactic| with_reducible
+    have := Pair.sizeOf_lt_of_mem ‹_›
+    omega
+    done)
+
+macro_rules | `(tactic| decreasing_trivial) => `(tactic| sizeOf_pair_dec)
+
+def Tree.map (f : α → β) : Tree α → Tree β
+  | leaf x => leaf (f x)
+  | node p => node (p.map (fun t' => t'.map f))
+termination_by t => t
+```
```

Comment:
Maybe a disclaimer like this doesn’t hurt:

````suggestion
```

Note that this this {tactic}`sizeOf_pair_dec` tactic is tailored to this particular recursion pattern, and not necessarily suited to be included in a general purpose container library.
````

Reply:
What do you think about the variation on this text that I just pushed?

Reply:
Perfect, ship it